src/crypto/hash160.cpp

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2018 jmjatlanta and contributors.
+ *
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <fc/crypto/hex.hpp>
+#include <fc/fwd_impl.hpp>
+#include <openssl/sha.h>
+#include <openssl/ripemd.h>
+#include <string.h>
+#include <fc/crypto/hash160.hpp>
+#include <fc/variant.hpp>
+#include <vector>
+#include "_digest_common.hpp"
+
+namespace fc 
+{
+  
+hash160::hash160() { memset( _hash, 0, sizeof(_hash) ); }
+
+hash160::hash160( const string& hex_str ) {
+   fc::from_hex( hex_str, (char*)_hash, sizeof(_hash) );  
+}
+
+string hash160::str()const {
+   return fc::to_hex( (char*)_hash, sizeof(_hash) );
+}
+
+hash160::operator string()const { return  str(); }
+
+char* hash160::data()const { return (char*)&_hash[0]; }
+
+class hash160::encoder::impl {
+   public:
+   impl() { }
+};
+
+hash160::encoder::~encoder() {}
+hash160::encoder::encoder() {}
+
+hash160 hash160::hash( const char* d, uint32_t dlen ) {
+   encoder e;
+   e.write(d,dlen);
+   return e.result();
+}
+
+hash160 hash160::hash( const string& s ) {
+   return hash( s.c_str(), s.size() );
+}
+
+void hash160::encoder::write( const char* d, uint32_t dlen )
+{
+   for(uint32_t i = 0; i < dlen; ++i)
+      bytes.push_back(d[i]); 
```

Comment:
Better reserve the space before the loop?  Please mind that `write()` can be called multiple times.

I think performance of this implementation (the whole class but not only this function) is not good enough especially when streaming or hashing large input data. Actually this can be easily improved since the streaming is SHA256 only. When calling `result()` we finalize the SHA256 streaming, then do the RIPEMD160 part which is small.

Reply:
Implemented as suggested.